components/base/kinematicbase/execution.go

```diff
@@ -0,0 +1,471 @@
+	}
+	// Trajectory distance is either length in mm, or if linear distance is not increasing, number of degrees to rotate in place.
+	lastLinVel := r3.Vector{0, traj[0].LinVel * ptgk.linVelocityMMPerSecond, 0}
+	lastAngVel := r3.Vector{0, 0, traj[0].AngVel * ptgk.angVelocityDegsPerSecond}
+	nextStep := arcStep{
+		linVelMMps:      lastLinVel,
+		angVelDegps:     lastAngVel,
+		arcSegment:      segment,
+		durationSeconds: 0.,
+	}
+	for _, trajPt := range traj {
+		nextStep.subTraj = append(nextStep.subTraj, trajPt)
+		nextLinVel := r3.Vector{0, trajPt.LinVel * ptgk.linVelocityMMPerSecond, 0}
+		nextAngVel := r3.Vector{0, 0, trajPt.AngVel * ptgk.angVelocityDegsPerSecond}
+
+		// Check if this traj node has different velocities from the last one. If so, end our segment and start a new segment.
+		if nextStep.linVelMMps.Sub(nextLinVel).Norm2() > 1e-6 || nextStep.angVelDegps.Sub(nextAngVel).Norm2() > 1e-6 {
+			// Changed velocity, make a new step
+			nextStep.durationSeconds = timeStep
+			curInputs = []referenceframe.Input{
+				inputs[0],
+				inputs[1],
+				{curDist},
+			}
+			arcPose, err := ptgk.Kinematics().Transform(curInputs)
+			if err != nil {
+				return nil, err
+			}
+			runningPose = spatialmath.Compose(runningPose, arcPose)
+			nextStep.arcSegment.EndConfiguration = curInputs
+			nextStep.arcSegment.EndPosition = runningPose
+			finalSteps = append(finalSteps, nextStep)
+			segment = ik.Segment{
+				StartConfiguration: curInputs,
+				StartPosition:      runningPose,
+				Frame:              ptgk.Kinematics(),
+			}
+			nextStep = arcStep{
+				linVelMMps:      nextLinVel,
+				angVelDegps:     nextAngVel,
+				arcSegment:      segment,
+				durationSeconds: 0,
+			}
+			timeStep = 0.
+		}
+		distIncrement := trajPt.Dist - curDist
+		curDist += distIncrement
+		if nextStep.linVelMMps.Y != 0 {
+			timeStep += distIncrement / (math.Abs(nextStep.linVelMMps.Y))
+		} else if nextStep.angVelDegps.Z != 0 {
+			timeStep += distIncrement / (math.Abs(nextStep.angVelDegps.Z))
+		}
+	}
+	nextStep.durationSeconds = timeStep
+	curInputs = []referenceframe.Input{
+		inputs[0],
+		inputs[1],
+		{curDist},
+	}
+	arcPose, err := ptgk.Kinematics().Transform(curInputs)
+	if err != nil {
+		return nil, err
+	}
+	runningPose = spatialmath.Compose(runningPose, arcPose)
+	nextStep.arcSegment.EndConfiguration = curInputs
+	nextStep.arcSegment.EndPosition = runningPose
+	finalSteps = append(finalSteps, nextStep)
+
+	return finalSteps, nil
+}
+
+func (ptgk *ptgBaseKinematics) courseCorrect(ctx context.Context, goals []courseCorrectionGoal) (courseCorrectionGoal, error) {
+	for _, goal := range goals {
+		solveMetric := ik.NewSquaredNormMetric(goal.Goal)
+		solutionChan := make(chan *ik.Solution, 1)
+		ptgk.logger.Debug("attempting goal ", spatialmath.PoseToProtobuf(goal.Goal))
+		seed := []referenceframe.Input{{math.Pi / 2}, {ptgk.linVelocityMMPerSecond / 2}, {math.Pi / 2}, {ptgk.linVelocityMMPerSecond / 2}}
+		if goal.Goal.Point().X > 0 {
```

Comment:
This section seems a little dubious to me and might be related to our spin issues

Reply:
And we only need to care about X since the goal is expressed with respect to the base frame?